zk/stages/stage_batches.go

```diff
@@ -231,149 +224,149 @@ LOOP:
-					panic(message)
+				// handle batch boundary changes - we do this here instead of reading the batch start channel because
+				// channels can be read in random orders which then creates problems in detecting fork changes during
+				// execution
+				if entry.BatchNumber > highestSeenBatchNo && lastForkId < entry.ForkId {
+					if entry.ForkId > HIGHEST_KNOWN_FORK {
+						message := fmt.Sprintf("unsupported fork id %v received from the data stream", entry.ForkId)
+						panic(message)
+					}
+					err = stages.SaveStageProgress(tx, stages.ForkId, entry.ForkId)
+					if err != nil {
+						return fmt.Errorf("save stage progress error: %v", err)
+					}
+					lastForkId = entry.ForkId
+					err = hermezDb.WriteForkId(entry.BatchNumber, entry.ForkId)
+					if err != nil {
+						return fmt.Errorf("write fork id error: %v", err)
+					}
+					// NOTE (RPC): avoided use of 'writeForkIdBlockOnce' by reading instead batch by forkId, and then lowest block number in batch
 				}
-				err = stages.SaveStageProgress(tx, stages.ForkId, l2Block.ForkId)
-				if err != nil {
-					return fmt.Errorf("save stage progress error: %v", err)
+
+				// ignore genesis or a repeat of the last block
+				if entry.L2BlockNumber == 0 {
+					continue
 				}
-				lastForkId = l2Block.ForkId
-				err = hermezDb.WriteForkId(l2Block.BatchNumber, l2Block.ForkId)
-				if err != nil {
-					return fmt.Errorf("write fork id error: %v", err)
+				// skip but warn on already processed blocks
+				if entry.L2BlockNumber <= stageProgressBlockNo {
+					if entry.L2BlockNumber < stageProgressBlockNo {
+						// only warn if the block is very old, we expect the very latest block to be requested
+						// when the stage is fired up for the first time
+						log.Warn(fmt.Sprintf("[%s] Skipping block %d, already processed", logPrefix, entry.L2BlockNumber))
+					}
+					continue
 				}
-				// NOTE (RPC): avoided use of 'writeForkIdBlockOnce' by reading instead batch by forkId, and then lowest block number in batch
-			}
 
-			// ignore genesis or a repeat of the last block
-			if l2Block.L2BlockNumber == 0 {
-				continue
-			}
-			// skip but warn on already processed blocks
-			if l2Block.L2BlockNumber <= stageProgressBlockNo {
-				if l2Block.L2BlockNumber < stageProgressBlockNo {
-					// only warn if the block is very old, we expect the very latest block to be requested
-					// when the stage is fired up for the first time
-					log.Warn(fmt.Sprintf("[%s] Skipping block %d, already processed", logPrefix, l2Block.L2BlockNumber))
+				// skip if we already have this block
+				if entry.L2BlockNumber < lastBlockHeight+1 {
+					log.Warn(fmt.Sprintf("[%s] Unwinding to block %d", logPrefix, entry.L2BlockNumber))
+					badBlock, err := eriDb.ReadCanonicalHash(entry.L2BlockNumber)
+					if err != nil {
+						return fmt.Errorf("failed to get bad block: %v", err)
+					}
+					u.UnwindTo(entry.L2BlockNumber, badBlock)
```

Comment:
when we unwind we actually don't reset what we request from the datastream (bookmark), so we are getting issues with "block number non sequential".

Reply:
see 
<img width="1039" alt="image" src="https://github.com/user-attachments/assets/804333ed-4e57-41c2-9d4f-8a863eeb8d6d">

Reply:
I don't think we should unwind in this case. This unwind doesn't restart the DS client. It just continues using the already downloaded blocks.

Reply:
The problem with this is upon disconnecting and then reconnecting. Then  we should reset the bookmark and probably do it to the latest block in the channel.